plane/src/drone/executor.rs

```diff
@@ -53,13 +61,77 @@ impl<R: Runtime> Executor<R> {
 
         Self {
             runtime,
-            state_store: Arc::new(Mutex::new(state_store)),
+            state_store,
             backends,
             ip,
             _backend_event_listener: backend_event_listener,
         }
     }
 
+    // On restart, we want to terminate all existing backends and start fresh.
+    // This prevents bugs where an agent restart leaves the drone unable to
+    // terminate old backends.
+    async fn terminate_preexisting_backends(
+        runtime: Arc<R>,
+        state_store: Arc<Mutex<StateStore>>,
+    ) -> Result<()> {
+        let backends = state_store
+            .lock()
+            .expect("State store lock poisoned.")
+            .active_backends()?;
+
+        tracing::info!(?backends, "Terminating preexisting backends");
+
+        let mut tasks = vec![];
+        let state_store_clone = state_store.clone();
+        for (backend_id, state) in backends {
+            let runtime_clone = runtime.clone();
+            let state_store = state_store_clone.clone();
+            let state = state.clone();
+            tasks.push(tokio::spawn(async move {
+                let mut backoff = ExponentialBackoff::default();
+                let mut success = false;
+                for attempt in 1..=10 {
+                    match runtime_clone.terminate(&backend_id, true).await {
+                        Ok(()) => {
+                            success = true;
+                            break;
+                        }
+                        Err(err) => {
+                            tracing::error!(
+                                ?err,
+                                ?backend_id,
+                                ?attempt,
+                                "Attempt failed to terminate backend"
+                            );
+                            backoff.wait().await;
+                        }
+                    }
+                }
+                if success {
+                    state_store
+                        .lock()
+                        .expect("State store lock poisoned.")
+                        .register_event(&backend_id, &state.to_terminated(Some(137)), Utc::now())
```

Comment:
I think we should call `to_terminating()` first with `TerminationKind::Hard` and `TerminationReason::KeyExpired`. Otherwise, those values won't be included in the final `Terminated` state.

Reply:
I'm not sure if we want to provide an exit code here that wasn't actually returned from the process. Do hard-terminated backends _always_ exit with a `137`?

Reply:
Also, do you know if these state transitions end up getting persisted to the sqlite db?

Reply:
If we don't have an exit code for a backend, we should use `None`; that's why it's an option (hard-terminated backends don't have any exit code by definition)

Reply:
Great feedback, addressing it all. Re: do these state transitions get persisted in the sqlite db -- yes, that's what  `state_store.reigster_event()` does.